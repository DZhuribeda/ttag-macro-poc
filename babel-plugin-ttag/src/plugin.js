import * as bt from '@babel/types';
import fs from 'fs';
import mkdirp from 'mkdirp';
import path from 'path';

import { ALIAS_TO_FUNC_MAP } from './defaults';
import { buildPotData, makePotStr } from './po-helpers';
import { extractPoEntry, getExtractor } from './extract';
import { hasDisablingComment, isInDisabledScope, isTtagImport,
    hasImportSpecifier, poReferenceComparator, isTtagRequire } from './utils';
import { resolveEntries } from './resolve';
import { ValidationError } from './errors';
import TtagContext from './context';
import { isContextTagCall, isValidTagContext, isContextFnCall,
    isValidFnCallContext } from './gettext-context';

export default function () {
    let context;
    let disabledScopes = new Set();
    const potEntries = [];

    function tryMatchTag(cb) {
        return (nodePath, state) => {
            const node = nodePath.node;
            if (isContextTagCall(node, context) && isValidTagContext(nodePath)) {
                nodePath._C3PO_GETTEXT_CONTEXT = node.tag.object.arguments[0].value;
                nodePath.node = bt.taggedTemplateExpression(node.tag.property, node.quasi);
                nodePath.node.loc = node.loc;
            }
            cb(nodePath, state);
        };
    }

    function tryMatchCall(cb) {
        return (nodePath, state) => {
            const node = nodePath.node;
            if (isContextFnCall(node, context) && isValidFnCallContext(nodePath)) {
                nodePath._C3PO_GETTEXT_CONTEXT = node.callee.object.arguments[0].value;
                nodePath.node = bt.callExpression(node.callee.property, node.arguments);
                nodePath.node.loc = node.loc;
            }
            cb(nodePath, state);
        };
    }

    function extractOrResolve(nodePath, state) {
        if (nodePath.node._C3PO_visited) { // Should visit each node only once
            return;
        }
        if (isInDisabledScope(nodePath, disabledScopes)) {
            return;
        }

        const extractor = getExtractor(nodePath, context);
        if (!extractor) {
            return;
        }

        const aliases = context.getAliasesForFunc(extractor.name);
        const hasImport = aliases.find(context.hasImport);
        if (!hasImport
            // can be used in scope of context without import
            && !nodePath._C3PO_GETTEXT_CONTEXT) {
            return;
        }

        try {
            try {
                extractor.validate(nodePath.node, context);
            } catch (err) {
                if (err instanceof ValidationError) {
                    context.validationFailureAction(extractor.name, err.message);
                    return;
                }
                throw err;
            }

            if (context.isExtractMode()) {
                const poEntry = extractPoEntry(extractor, nodePath, context, state);
                poEntry && potEntries.push(poEntry);
            }

            if (context.isResolveMode()) {
                resolveEntries(extractor, nodePath, context, state);
            }
            nodePath.node._C3PO_visited = true;
        } catch (err) {
            // TODO: handle specific instances of errors
            throw nodePath.buildCodeFrameError(`${err.message}\n${err.stack}`);
        }
    }

    return {
        post() {
            if (context && context.isExtractMode() && potEntries.length) {
                const poData = buildPotData(potEntries);

                // Here we sort reference entries, this could be useful
                // with conf. options extract.location: 'file' and sortByMsgid
                // which simplifies merge of .po files from different
                // branches of SCM such as git or mercurial.
                const ctxs = Object.keys(poData.translations);
                for (const ctx of ctxs) {
                    const poEntries = poData.translations[ctx];
                    Object.keys(poEntries).forEach((k) => {
                        const poEntry = poEntries[k];
                        // poEntry has a form:
                        // {
                        //     msgid: 'message identifier',
                        //     msgstr: 'translation string',
                        //     comments: {
                        //         reference: 'path/to/file.js:line_number\npath/to/other/file.js:line_number'
                        //     }
                        // }
                        if (poEntry.comments && poEntry.comments.reference) {
                            poEntry.comments.reference = poEntry.comments.reference
                                .split('\n')
                                .sort(poReferenceComparator)
                                .join('\n');
                        }
                    });

                    if (context.isSortedByMsgid()) {
                        const oldPoData = poData.translations[ctx];
                        const newContext = {};
                        const keys = Object.keys(oldPoData).sort();
                        keys.forEach((k) => {
                            newContext[k] = oldPoData[k];
                        });
                        poData.translations[ctx] = newContext;
                    }
                }
                const potStr = makePotStr(poData);
                const filepath = context.getOutputFilepath();
                const dirPath = path.dirname(filepath);
                mkdirp.sync(dirPath);
                fs.writeFileSync(filepath, potStr);
            }
        },
        visitor: {
            TaggedTemplateExpression: tryMatchTag(extractOrResolve),
            CallExpression: tryMatchCall(extractOrResolve),
            Program: (nodePath, state) => {
                if (!context) {
                    context = new TtagContext(state.opts);
                } else {
                    context.clear();
                }
                disabledScopes = new Set();
                if (hasDisablingComment(nodePath.node)) {
                    disabledScopes.add(nodePath.scope.uid);
                }
            },
            BlockStatement: (nodePath) => {
                if (hasDisablingComment(nodePath.node)) {
                    disabledScopes.add(nodePath.scope.uid);
                }
            },
            VariableDeclarator: (nodePath) => {
                const { node } = nodePath;
                if (!isTtagRequire(node)) return;

                // require calls
                node.id.properties
                    .map(({ key: { name: keyName }, value: { name: valueName } }) => [keyName, valueName])
                    .filter(([keyName]) => ALIAS_TO_FUNC_MAP[keyName])
                    .forEach(([keyName, valueName]) => {
                        if (keyName !== valueName) { // if alias
                            context.addAlias(ALIAS_TO_FUNC_MAP[keyName], valueName);
                            context.addImport(valueName);
                        } else {
                            context.addImport(keyName);
                        }
                    });
            },
            ImportDeclaration: (nodePath, state) => {
                const { node } = nodePath;

                if (!context) {
                    context = new TtagContext(state.opts);
                }
                if (isTtagImport(node)) {
                    node.specifiers
                    .filter(({ local: { name } }) => ALIAS_TO_FUNC_MAP[name])
                    .forEach((s) => context.addImport(s.local.name));
                }
                if (isTtagImport(node) && hasImportSpecifier(node)) {
                    node.specifiers
                    .filter(bt.isImportSpecifier)
                    .filter(({ imported: { name } }) => ALIAS_TO_FUNC_MAP[name])
                    .forEach(({ imported, local }) => {
                        context.addAlias(ALIAS_TO_FUNC_MAP[imported.name], local.name);
                        context.addImport(local.name);
                    });
                }
            },
        },
    };
}
