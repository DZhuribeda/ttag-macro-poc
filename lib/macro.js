import { createMacro, MacroError } from 'babel-plugin-macros';
import babelPluginTtag from 'babel-plugin-ttag';
import * as ttag from 'ttag';

const allowedImports = Object.keys(ttag).filter(helper => helper !== '__esModule');

function ttagMacro({ references, state, babel: { types: t }, config = {} }) {
  const program = state.file.path;

  // replace `ttag.macro` by `ttag`
  // create a node for ttag's imports
  const imports = t.importDeclaration([], t.stringLiteral('ttag'));
  // and add it to top of the document
  program.node.body.unshift(imports);

  // references looks like this
  // { default: [path, path], css: [path], ... }
  Object.keys(references).forEach(refName => {
    if (!allowedImports.includes(refName)) {
      throw new MacroError(
        `Invalid import: ${refName}. You can only import ${allowedImports.join(
          ', '
        )} from 'ttag.macro'.`
      );
    }

    // generate new identifier and add to imports
    let id;
    if (refName === 'default') {
      id = program.scope.generateUidIdentifier('ttag');
      imports.specifiers.push(t.importDefaultSpecifier(id));
    } else {
      id = program.scope.generateUidIdentifier(refName);
      imports.specifiers.push(t.importSpecifier(id, t.identifier(refName)));
    }

    // update references with the new identifiers
    references[refName].forEach(referencePath => {
      // eslint-disable-next-line no-param-reassign
      referencePath.node.name = id.name;
    });
  });

  // apply babel-plugin-ttag to the file
  const stateWithOpts = { ...state, opts: config };
  program.traverse(babelPluginTtag().visitor, stateWithOpts);
}

export default createMacro(ttagMacro, { configName: 'ttag' });
