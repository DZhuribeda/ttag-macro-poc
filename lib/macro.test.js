import cosmiconfigMock from 'cosmiconfig';
import pluginTester from 'babel-plugin-tester';
import plugin from 'babel-plugin-macros';

jest.mock('cosmiconfig', () => jest.fn(require.requireActual('cosmiconfig')));

const exampleCode = `
import { t, jt, gettext, ngettext, msgid } from './macro';

const a = 'text';
const b = ['text 1', 'text 2'];

function pluralText(n) {
  return ngettext(
    msgid\`plural form with \${ n } plural\`,
    \`plural form with \${ n } plurals\`,
    n
  );
}

console.log(t\`simple string literal\`);
console.log(jt\`\${ a } simple string \${ b } literal with formatting\`);
console.log(t\`no translator notes\`);
console.log(pluralText(1));
console.log(pluralText(2));
console.log(t\`
  first line
  second line
  third line
\`);
console.log(gettext\`\${ a } spaces test\`);
console.log(t\`{name} fuzzy name\`);
`;

const errorExampleCode = `
import { t } from './macro';

const mississipi = 'test';

console.log(t\`Typo test \${ mississipi }\`);
`;

pluginTester({
  title: 'macro',
  plugin,
  snapshot: true,
  babelOptions: { filename: __filename },
  tests: {
    'should replace translations': {
      code: exampleCode,
      setup: () => {
        cosmiconfigMock.mockImplementationOnce(() => ({
          searchSync: () => ({
            config: {
              ttag: {
                resolve: { translations: 'translations.po' },
              },
            },
          }),
        }));
      },
    },
    'should throw error if missing placeholder': {
      code: errorExampleCode,
      error: /Expression 'mississipi' is not found in the localized string 'Typo test \${ missingpi }'./,
      snapshot: false,
      setup: () => {
        cosmiconfigMock.mockImplementationOnce(() => ({
          searchSync: () => ({
            config: {
              ttag: {
                resolve: { translations: 'translations.po' },
              },
            },
          }),
        }));
      },
    },
  },
});
