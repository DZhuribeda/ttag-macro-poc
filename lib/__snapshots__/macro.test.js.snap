// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`macro should replace translations: should replace translations 1`] = `
"
import { t, jt, gettext, ngettext, msgid } from './macro';

const a = 'text';
const b = ['text 1', 'text 2'];

function pluralText(n) {
  return ngettext(
    msgid\`plural form with \${ n } plural\`,
    \`plural form with \${ n } plurals\`,
    n
  );
}

console.log(t\`simple string literal\`);
console.log(jt\`\${ a } simple string \${ b } literal with formatting\`);
console.log(t\`no translator notes\`);
console.log(pluralText(1));
console.log(pluralText(2));
console.log(t\`
  first line
  second line
  third line
\`);
console.log(gettext\`\${ a } spaces test\`);
console.log(t\`{name} fuzzy name\`);

      ↓ ↓ ↓ ↓ ↓ ↓

function _tag_ngettext(n, args) {
  return args[+(n != 1)];
}

import { t as _t, jt as _jt, gettext as _gettext, ngettext as _ngettext, msgid as _msgid } from \\"ttag\\";
const a = 'text';
const b = ['text 1', 'text 2'];

function pluralText(n) {
  return _tag_ngettext(n, [\`plural form with \${n} plural [translated]\`, \`plural form with \${n} plurals [translated]\`]);
}

console.log(\\"simple string literal translated\\");
console.log([a, \\" simple string \\", b, \\" literal with formatting [translated]\\"]);
console.log(\\"no translator notes\\");
console.log(pluralText(1));
console.log(pluralText(2));
console.log(\\"translation\\");
console.log(_gettext\`\${a} spaces test\`);
console.log(\\"{name} fuzzy name\\");
"
`;
